### ISSUE TITLE
New cmake modules commit breaks android build

### ISSUE THREAD

**Issue by spiderkeys** [eProsima/Fast-RTPS#80]
Commit: 67e42c8a52ae631ad198f3da20ceb782fbb6f6bf

CMake fails to find the ASIO package when building from android. This may have been a case of the development machine already having the packages installed, such that find_package worked. On a fresh build, the find fails.

Input:
```sh
git clone --recursive https://github.com/eProsima/Fast-RTPS
cd Fast-RTPS

mkdir build
cd build

export CMAKE_EXECUTABLE=$HOME/Android/Sdk/cmake/3.6.3155560/bin/cmake
export TOOLCHAIN_PATH=$HOME/android/android-ndk-r14/build/cmake/android.toolchain.cmake
export NDK_PATH=$HOME/Android/Sdk/ndk-bundle

${CMAKE_EXECUTABLE} ../ \
-DCMAKE_INSTALL_PREFIX=install \
-DCMAKE_TOOLCHAIN_FILE=${TOOLCHAIN_PATH} \
-DCMAKE_BUILD_TYPE=Release \
-DEPROSIMA_BUILD=ON \
-DANDROID_NDK=${NDK_PATH} \
-DANDROID_TOOLCHAIN_NAME=arm-linux-androideabi-clang \
-DANDROID_ABI=armeabi-v7a \
-DANDROID_NATIVE_API_LEVEL=android-21 \
-DANDROID_PLATFORM=android-21 \
-DANDROID_FUNCTION_LEVEL_LINKING=OFF \
-DANDROID_STL=gnustl_shared \
-DCOMPILE_EXAMPLES=OFF
```

Trailing end of output:
```sh
...
...
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- fastcdr library not found...
CMake Error at /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find asio (missing: ASIO_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/Findasio.cmake:7 (find_package_handle_standard_args)
  cmake/dev/eprosima_libraries.cmake:139 (find_package)
  CMakeLists.txt:195 (find_eprosima_thirdparty)
```

The cmake incantation in the Readme works for Linux, and my above incantation works for commit 5301ef203d45528a083821c3ba582164d782360b.

**Comment by spiderkeys** [comment]
The following produces a successful build:

```sh
git clone https://github.com/eProsima/Fast-RTPS.git
cd Fast-RTPS
git reset --hard 4fcd335865ad426a223494f407736330ef94d365
git submodule update --init --recursive

export CMAKE_EXECUTABLE=$HOME/Android/Sdk/cmake/3.6.3155560/bin/cmake
export TOOLCHAIN_PATH=$HOME/android/android-ndk-r14/build/cmake/android.toolchain.cmake
export NDK_PATH=$HOME/Android/Sdk/ndk-bundle

mkdir build
cd build

${CMAKE_EXECUTABLE} ../ \
-DCMAKE_INSTALL_PREFIX=install \
-DCMAKE_TOOLCHAIN_FILE=${TOOLCHAIN_PATH} \
-DCMAKE_BUILD_TYPE=Release \
-DEPROSIMA_BUILD=ON \
-DANDROID_NDK=${NDK_PATH} \
-DANDROID_TOOLCHAIN_NAME=arm-linux-androideabi-clang \
-DANDROID_ABI=armeabi-v7a \
-DANDROID_NATIVE_API_LEVEL=android-21 \
-DANDROID_PLATFORM=android-21 \
-DANDROID_FUNCTION_LEVEL_LINKING=OFF \
-DANDROID_STL=gnustl_shared \
-DCOMPILE_EXAMPLES=OFF

make -j
```

**Comment by spiderkeys** [comment]
Interestingly, fastcdr gets built for the wrong architecture when using the above:

```sh
readelf -h ./external/install/lib/libfastcdr.so
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x53a0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          87512 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         6
  Size of section headers:           64 (bytes)
  Number of section headers:         30
  Section header string table index: 27
```

Building fastcdr on its own produces the correct information:
```sh
~/buildtest/Fast-RTPS/thirdparty/fastcdr/build$ readelf -h ./src/cpp/libfastcdr.so
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           ARM
  Version:                           0x1
  Entry point address:               0x0
  Start of program headers:          52 (bytes into file)
  Start of section headers:          676392 (bytes into file)
  Flags:                             0x5000200, Version5 EABI, soft-float ABI
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         8
  Size of section headers:           40 (bytes)
  Number of section headers:         40
  Section header string table index: 39
```

**Comment by richiware** [comment]
eProsima/Fast-RTPS@634f38f51561ad4a0c181d71fd2d5a75f499c103 and eProsima/Fast-RTPS@7c91865448e734355192809e97a94b426148574a should solve your problem. I could compile Fast-RTPS for Android. My command line was:

```
cmake -DCMAKE_INSTALL_PREFIX=install \
           -DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk/build/cmake/android.toolchain.cmake \
           -DANDROID_NDK=/opt/android-ndk \
           -DANDROID_ABI=armeabi-v7a \
           -DANDROID_TOOLCHAIN=gcc \
           -DTHIRDPARTY=ON \
           -DCMAKE_CXX_FLAGS="-frtti -fexceptions" \
           -DANDROID_NATIVE_API_LEVEL=android-21 ..
```
Also now when FastRTPS compiles FastCDR, it pass all android options to FastCDR compilation.

**Comment by spiderkeys** [comment]
Thanks @richiware, those commits sorted it out. I've added PR #81, which adds the rest of the variables listed at https://developer.android.com/ndk/guides/cmake.html, as well as the function level linking option that @esteve specified in his ROS2 Java build instructions. This was mainly to make sure that the ANDROID_STL option made its way through.

Additionally, I have written a short script that builds a distribution of FastRTPS for all of the Android ABIs listed on https://developer.android.com/ndk/guides/abis.html, except armeabi (some kind of build errors there). Do you think something like this might be useful to include somewhere as a guide for helping people get started on Android?

```sh
set -xe

git clone --recursive https://github.com/eProsima/Fast-RTPS.git
cd Fast-RTPS

# Set home dir
export DIR=$PWD

# Create dist directories
mkdir -p dist
mkdir -p dist/include
mkdir -p dist/lib

# Set tool paths
SDK_PATH=$HOME/Android/Sdk
NDK_PATH=$SDK_PATH/ndk-bundle
CMAKE_EXECUTABLE=$SDK_PATH/cmake/3.6.3155560/bin/cmake
TOOLCHAIN_PATH=$NDK_PATH/build/cmake/android.toolchain.cmake

# Declare ABIs to build for
declare -a arr=( "armeabi-v7a" "arm64-v8a" "x86" "x86_64" "mips" "mips64" )

for ABI in "${arr[@]}"
do
	# Create ABI specific cmake build dir
	mkdir -p build_${ABI}

	# Create ABI specific output directory
	mkdir -p $DIR/dist/lib/${ABI}/

	# Move into ABI specific build dir
	cd build_${ABI}

	# Build makefile
	${CMAKE_EXECUTABLE} \
	-DCMAKE_INSTALL_PREFIX=install \
	-DCMAKE_TOOLCHAIN_FILE=${TOOLCHAIN_PATH} \
	-DCMAKE_BUILD_TYPE=Release \
	-DANDROID_NDK=${NDK_PATH} \
	-DANDROID_TOOLCHAIN=clang \
	-DANDROID_ABI=${ABI} \
	-DANDROID_NATIVE_API_LEVEL=android-21 \
	-DANDROID_STL=gnustl_shared \
	-DCMAKE_CXX_FLAGS="-frtti -fexceptions" \
	-DTHIRDPARTY=ON \
	-DCOMPILE_EXAMPLES=OFF \
	${DIR}

	# Build and install
	make -j
	make install

	# Copy artifacts to 
	cp install/lib/libfastrtps.so $DIR/dist/lib/${ABI}/
	cp external/install/lib/libfastcdr.so $DIR/dist/lib/${ABI}/

	# Move back to main directory
	cd $DIR
done

# Copy include folders
cp -r include/fastrtps dist/include
cp -r thirdparty/fastcdr/include/fastcdr dist/include

# Zip
FASTRTPS_HASH=$(git rev-parse --short HEAD)
zip -r libfastrtps_${FASTRTPS_HASH}.zip dist/*
```

**Comment by spiderkeys** [comment]
Fixed in #22.
